### PR TITLE
drivers: wifi: Store the negotiated TWT flow id at host level.

### PR DESCRIPTION
**Pull request by ajayparida**
[NRF7X-17] Store the negotaited twt flow id at host level, for later use to tear down all TWT sessions.

Signed-off-by: Ajay Parida <ajay.parida@nordicsemi.no>

### CODE REVIEW

File: drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h [hunk 1]
```diff
@@ -47,6 +47,10 @@ struct wifi_nrf_vif_ctx_zep {
 #ifdef CONFIG_WPA_SUPP
 	struct zep_wpa_supp_dev_callbk_fns supp_callbk_fns;
 #endif /* CONFIG_WPA_SUPP */
+	/* Used to store the negotaited twt flow id
+	 * for "twt_release_all" command.
+	 */
+	unsigned char negotiated_twt_flow_id;
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h:53]: Use short forms where possible (e.g. in this case negotiated can be just neg)

File: drivers/wifi/nrf700x/zephyr/src/zephyr_twt.c [hunk 2]
```diff
@@ -74,6 +74,14 @@ int wifi_nrf_set_twt(const struct device *dev,
 	case WIFI_TWT_TEARDOWN:
 		twt_info.twt_flow_id = twt_params->flow_id;
 
+		if (twt_params->teardown.teardown_all) {
+			/* update the negotiated flow id from the negotaited value. */
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_twt.c:78]: Use capital letter for the first letter in a sentence (i.e. update should be Update)
Also check for spelling mistakes (in this case negotaited should be negotiated)

File: drivers/wifi/nrf700x/zephyr/src/zephyr_twt.c [hunk 3]
```diff
@@ -74,6 +74,14 @@ int wifi_nrf_set_twt(const struct device *dev,
...
+			twt_info.twt_flow_id = vif_ctx_zep->negotiated_twt_flow_id;
+		} else {
+			/* update the flow id as given by the user. */
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_twt.c:81]: Use capital letter for the first letter in a sentence (i.e. update should be Update)

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

File: drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h [hunk 4]
```diff
@@ -47,6 +47,10 @@ struct wifi_nrf_vif_ctx_zep {
 #ifdef CONFIG_WPA_SUPP
 	struct zep_wpa_supp_dev_callbk_fns supp_callbk_fns;
 #endif /* CONFIG_WPA_SUPP */
+	/* Used to store the negotaited twt flow id
+	 * for "twt_release_all" command.
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h:51]: I think the command is `twt_reardown_all`

File: drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h [hunk 5]
```diff
@@ -47,6 +47,10 @@ struct wifi_nrf_vif_ctx_zep {
...
+	 */
+	unsigned char neg_twt_flow_id;
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h:53]: Just for future proofing, should we store make this an array?

File: drivers/wifi/nrf700x/zephyr/src/zephyr_twt.c [hunk 6]
```diff
@@ -58,7 +58,7 @@ int wifi_nrf_set_twt(const struct device *dev,
 		twt_info.ap_trigger_frame = twt_params->setup.trigger;
 		twt_info.is_implicit = twt_params->setup.implicit;
 		twt_info.twt_flow_type = twt_params->setup.announce;
-		/* interval values are passed as us to UMAC */
+		/* Interval values are passed as us to UMAC */
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_twt.c:61]: unrelated change

File: drivers/wifi/nrf700x/zephyr/src/zephyr_twt.c [hunk 7]
```diff
@@ -114,6 +122,8 @@ void wifi_nrf_event_proc_twt_setup_zep(void *vif_ctx,
 
 	twt_params.operation = WIFI_TWT_SETUP;
 	twt_params.flow_id = twt_setup_info->info.twt_flow_id;
+	/* Store the negotiated flow id and pass it to user. */
+	vif_ctx_zep->neg_twt_flow_id = twt_setup_info->info.twt_flow_id;
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_twt.c:126]: Where is this being cleared? This should be cleared in both all/specific teardown.

**Review by krish2718 — CHANGES_REQUESTED** [review]
I would recommend maintaining full array of flow IDs and then looping through them, even though we support only 1 flow for now.

File: drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h [hunk 8]
```diff
@@ -47,6 +47,10 @@ struct wifi_nrf_vif_ctx_zep {
 #ifdef CONFIG_WPA_SUPP
 	struct zep_wpa_supp_dev_callbk_fns supp_callbk_fns;
 #endif /* CONFIG_WPA_SUPP */
+	/* Used to store the negotaited twt flow id
```
Comment (sachinthegreen) [inline drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h:50]: negotaited should be negotiated

**Review by sachinthegreen — CHANGES_REQUESTED** [review]

**Review by sr1dh48r — APPROVED** [review]

File: drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h (hunk 5 above)
Comment (ajayparida) [reply drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h:53]: This is not part of the current release, only one flow id is being considered.

Comment (krish2718) [reply drivers/wifi/nrf700x/zephyr/inc/zephyr_fmac_main.h:53]: Sure, that is why I mentioned future proofing, its simple to implement as well.

File: drivers/wifi/nrf700x/zephyr/src/zephyr_twt.c [hunk 9]
```diff
@@ -74,9 +74,22 @@ int wifi_nrf_set_twt(const struct device *dev,
 	case WIFI_TWT_TEARDOWN:
 		twt_info.twt_flow_id = twt_params->flow_id;
 
+		if (twt_params->teardown.teardown_all) {
+			/* Update the negotiated flow id from the negotiated value. */
+			twt_info.twt_flow_id = vif_ctx_zep->neg_twt_flow_id;
```
Comment (krish2718) [inline drivers/wifi/nrf700x/zephyr/src/zephyr_twt.c:79]: Shouldn't we skip sending teardown_all if the flow_id == 0XFF?
